### PR TITLE
Fix incorrect column mapping in Kudu connector

### LINKED ISSUE #2963: Incorrect Kudu predicate pushdowm
```
1.create table kudu.test.test1
CREATE TABLE kudu.test.test2test1 (
   finishedtime timestamp WITH ( primary_key = true ),
   uid bigint WITH ( primary_key = true ),
   order_no varchar WITH ( primary_key = true ),
   id bigint WITH ( nullable = true ),
   total_fee decimal(10,2) WITH ( nullable = true ),
   verifytime timestamp WITH ( nullable = true ),
   reg_time timestamp WITH ( nullable = true )
)
WITH (
   number_of_replicas = 1,
   partition_by_hash_buckets = 3,
   partition_by_hash_columns = ARRAY['uid'],
   partition_by_range_columns = ARRAY['finishedtime'],
   range_partitions = '[{"lower":"2018-01-01T00:00:00.000Z","upper":"2021-01-01T00:00:00.000Z"}]'
)
​
2.create table kudu.test.test2
CREATE TABLE kudu.test.test2 (
   order_no varchar WITH ( primary_key = true ),
   verifytime timestamp WITH ( primary_key = true )
)
WITH (
   number_of_replicas = 1,
   partition_by_hash_buckets = 3,
   partition_by_hash_columns = ARRAY['order_no'],
   partition_by_range_columns = ARRAY['verifytime'],
   range_partitions = '[{"lower":"2018-01-01T00:00:00.000Z","upper":"2021-01-01T00:00:00.000Z"}]'
)
​
3.insert data into table  
insert into kudu.test.test1 values (timestamp '2020-02-25 08:00:01',1,'a1',1,125.2,timestamp '2020-02-25 08:20:00',timestamp '2020-02-20 08:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:05:42',2,'a2',2,180.2,timestamp '2020-02-25 08:23:00',timestamp '2020-02-20 09:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:12:33',3,'a3',3,550.2,timestamp '2020-02-25 08:32:00',timestamp '2020-02-20 10:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:22:55',4,'b1',4,150.2,timestamp '2020-02-25 08:41:00',timestamp '2020-02-20 11:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:31:12',5,'b2',5,120.2,timestamp '2020-02-25 08:49:00',timestamp '2020-02-20 12:00:00');
​
insert into kudu.test.test2 values ('a1',timestamp '2020-02-25 08:01:12');
insert into kudu.test.test2 values ('a2',timestamp '2020-02-25 08:21:22');
insert into kudu.test.test2 values ('b6',timestamp '2020-02-25 08:32:32');
insert into kudu.test.test2 values ('b7',timestamp '2020-02-25 08:33:12');
insert into kudu.test.test2 values ('c8',timestamp '2020-02-25 08:34:11');
​
4.use test sql in different version of presto
select uid,order_no,total_fee as total_fee1,total_fee as total_fee2,finishedtime as finishedtime1,finishedtime as finishedtime2,date_diff('day',reg_time,finishedtime)+1 as day_dif,reg_time
from kudu.test.test1 
where finishedtime > timestamp '2020-02-25 08:00:00' and finishedtime <= timestamp '2020-02-25 08:20:00'
and order_no not in 
(select order_no 
from kudu.test.test2
where verifytime > timestamp '2020-02-25 08:00:00' and verifytime <= timestamp '2020-02-25 08:20:00')  ; 
​
5.result  
   5.1 version of 322   -- only succeed
    uid | order_no | total_fee1 | total_fee2 |      finishedtime1      |      finishedtime2      | day_dif |        reg_time         
-----+----------+------------+------------+-------------------------+-------------------------+---------+-------------------------
   2 | a2       |     180.20 |     180.20 | 2020-02-25 08:05:42.000 | 2020-02-25 08:05:42.000 |       5 | 2020-02-20 09:00:00.000 
   3 | a3       |     550.20 |     550.20 | 2020-02-25 08:12:33.000 | 2020-02-25 08:12:33.000 |       5 | 2020-02-20 10:00:00.000
   
   5.2 version of 328   --failed
   Query 20200228_021532_00005_s4zi7 failed: Column (name: id, index: 3) is of type int64 but was requested as a type [Type: decimal]
   
   5.3 version of 330   --failed
   Query 20200228_021532_00005_s4zi7 failed: Column (name: id, index: 3) is of type int64 but was requested as a type [Type: decimal]
Collapse
```

Other test:
```
1.create table 
description
--The only differences between test1 and test3 are the type of total_fee. In test1 total_fee is decimal, In test3 total_fee is bigint(The same type with id and its index is just behind id)
CREATE TABLE kudu.test.test1 (
   finishedtime timestamp WITH ( primary_key = true ),
   uid bigint WITH ( primary_key = true ),
   order_no varchar WITH ( primary_key = true ),
   id bigint WITH ( nullable = true ),
   total_fee decimal(10,2) WITH ( nullable = true ),
   verifytime timestamp WITH ( nullable = true ),
   reg_time timestamp WITH ( nullable = true )
)
WITH (
   number_of_replicas = 1,
   partition_by_hash_buckets = 3,
   partition_by_hash_columns = ARRAY['uid'],
   partition_by_range_columns = ARRAY['finishedtime'],
   range_partitions = '[{"lower":"2018-01-01T00:00:00.000Z","upper":"2021-01-01T00:00:00.000Z"}]'
)
​
​
CREATE TABLE kudu.test.test2 (
   order_no varchar WITH ( primary_key = true ),
   verifytime timestamp WITH ( primary_key = true )
)
WITH (
   number_of_replicas = 1,
   partition_by_hash_buckets = 3,
   partition_by_hash_columns = ARRAY['order_no'],
   partition_by_range_columns = ARRAY['verifytime'],
   range_partitions = '[{"lower":"2018-01-01T00:00:00.000Z","upper":"2021-01-01T00:00:00.000Z"}]'
)
​
​
CREATE TABLE kudu.test.test3(
   finishedtime timestamp WITH ( primary_key = true ),
   uid bigint WITH ( primary_key = true ),
   order_no varchar WITH ( primary_key = true ),
   id bigint WITH ( nullable = true ),
   total_fee bigint WITH ( nullable = true ),
   verifytime timestamp WITH ( nullable = true ),
   reg_time timestamp WITH ( nullable = true )
)
WITH (
   number_of_replicas = 1,
   partition_by_hash_buckets = 3,
   partition_by_hash_columns = ARRAY['uid'],
   partition_by_range_columns = ARRAY['finishedtime'],
   range_partitions = '[{"lower":"2018-01-01T00:00:00.000Z","upper":"2021-01-01T00:00:00.000Z"}]'
)
​
​
2.insert test data 
description
--The only differences between test1 and test3 are the data of total_fee. 
insert into kudu.test.test1 values (timestamp '2020-02-25 08:00:01',1,'a1',1,125.2,timestamp '2020-02-25 08:20:00',timestamp '2020-02-20 08:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:05:42',2,'a2',2,180.2,timestamp '2020-02-25 08:23:00',timestamp '2020-02-20 09:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:12:33',3,'a3',3,550.2,timestamp '2020-02-25 08:32:00',timestamp '2020-02-20 10:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:22:55',4,'b1',4,150.2,timestamp '2020-02-25 08:41:00',timestamp '2020-02-20 11:00:00');
insert into kudu.test.test1 values (timestamp '2020-02-25 08:31:12',5,'b2',5,120.2,timestamp '2020-02-25 08:49:00',timestamp '2020-02-20 12:00:00');
​
insert into kudu.test.test2 values ('a1',timestamp '2020-02-25 08:01:12');
insert into kudu.test.test2 values ('a2',timestamp '2020-02-25 08:21:22');
insert into kudu.test.test2 values ('b6',timestamp '2020-02-25 08:32:32');
insert into kudu.test.test2 values ('b7',timestamp '2020-02-25 08:33:12');
insert into kudu.test.test2 values ('c8',timestamp '2020-02-25 08:34:11');
​
insert into kudu.test.test3 values (timestamp '2020-02-25 08:00:01',1,'a1',1,125,timestamp '2020-02-25 08:20:00',timestamp '2020-02-20 08:00:00');
insert into kudu.test.test3 values (timestamp '2020-02-25 08:05:42',2,'a2',2,180,timestamp '2020-02-25 08:23:00',timestamp '2020-02-20 09:00:00');
insert into kudu.test.test3 values (timestamp '2020-02-25 08:12:33',3,'a3',3,550,timestamp '2020-02-25 08:32:00',timestamp '2020-02-20 10:00:00');
insert into kudu.test.test3 values (timestamp '2020-02-25 08:22:55',4,'b1',4,150,timestamp '2020-02-25 08:41:00',timestamp '2020-02-20 11:00:00');
insert into kudu.test.test3 values (timestamp '2020-02-25 08:31:12',5,'b2',5,120,timestamp '2020-02-25 08:49:00',timestamp '2020-02-20 12:00:00');
​
4.to do test
description
--The only difference between sql_1 and sql_2 is the table. 
​
sql_1    test1 and test2
select uid,order_no,total_fee as total_fee1,total_fee as total_fee2,finishedtime as finishedtime1,finishedtime as finishedtime2,date_diff('day',reg_time,finishedtime)+1 as day_dif,reg_time
from kudu.test.test1 
where finishedtime > timestamp '2020-02-25 08:00:00' and finishedtime <= timestamp '2020-02-25 08:20:00'
and order_no not in 
(select order_no 
from kudu.test.test2
where verifytime > timestamp '2020-02-25 08:00:00' and verifytime <= timestamp '2020-02-25 08:20:00')  ; 
​
sql_2  test3 and test2 
select uid,order_no,total_fee as total_fee1,total_fee as total_fee2,finishedtime as finishedtime1,finishedtime as finishedtime2,date_diff('day',reg_time,finishedtime)+1 as day_dif,reg_time
from kudu.test.test3 
where finishedtime > timestamp '2020-02-25 08:00:00' and finishedtime <= timestamp '2020-02-25 08:20:00'
and order_no not in 
(select order_no 
from kudu.test.test2
where verifytime > timestamp '2020-02-25 08:00:00' and verifytime <= timestamp '2020-02-25 08:20:00')  ; 
​
5.result  
 sql_1
   5.1.1 version of 322   -- only succeed and right
   
 uid | order_no | total_fee1 | total_fee2 |      finishedtime1      |      finishedtime2      | day_dif |        reg_time         
-----+----------+------------+------------+-------------------------+-------------------------+---------+-------------------------
   2 | a2       |     180.20 |     180.20 | 2020-02-25 08:05:42.000 | 2020-02-25 08:05:42.000 |       5 | 2020-02-20 09:00:00.000 
   3 | a3       |     550.20 |     550.20 | 2020-02-25 08:12:33.000 | 2020-02-25 08:12:33.000 |       5 | 2020-02-20 10:00:00.000
   
   5.1.2 version of 328   --failed
   Query 20200228_021532_00005_s4zi7 failed: Column (name: id, index: 3) is of type int64 but was requested as a type [Type: decimal]
   
   5.1.3 version of 330   --failed
   Query 20200228_021532_00005_s4zi7 failed: Column (name: id, index: 3) is of type int64 but was requested as a type [Type: decimal]
 
 
 sql_2
	All are succeed ---But, i think, The reason comes up
	
	5.2.1  result    version of 322     -- right result
 uid | order_no | total_fee1 | total_fee2 |      finishedtime1      |      finishedtime2      | day_dif |        reg_time         
-----+----------+------------+------------+-------------------------+-------------------------+---------+-------------------------
   2 | a2       |        180 |        180 | 2020-02-25 08:05:42.000 | 2020-02-25 08:05:42.000 |       5 | 2020-02-20 09:00:00.000 
   3 | a3       |        550 |        550 | 2020-02-25 08:12:33.000 | 2020-02-25 08:12:33.000 |       5 | 2020-02-20 10:00:00.000	
    
	5.2.2 result    version of 328 and 330     -- wrong results
 uid | order_no | total_fee1 | total_fee2 |      finishedtime1      |      finishedtime2      | day_dif |        reg_time         
-----+----------+------------+------------+-------------------------+-------------------------+---------+-------------------------
   2 | a2       |          2 |          2 | 2020-02-25 08:05:42.000 | 2020-02-25 08:05:42.000 |   18318 | 1970-01-01 08:00:00.000 
   3 | a3       |          3 |          3 | 2020-02-25 08:12:33.000 | 2020-02-25 08:12:33.000 |   18318 | 1970-01-01 08:00:00.000
```
--conclusion
You can see the column of total_fee1, day_dif and reg_time in sql_2's result
I think the reason is the index that you have been used by mistake in #2253

### LINKED ISSUE #3170: Result fields are misaligned when using subqueries with Kudu connector
--- this is a fact table
 CREATE TABLE kudu.test.test_fact (                       
    id varchar WITH ( primary_key = true ),    
    grade varchar WITH ( primary_key = true ),
    score varchar WITH ( nullable = true ),
    name varchar WITH ( nullable = true ),
    age varchar WITH ( nullable = true ),
    school varchar WITH ( nullable = true ),
	teacher varchar WITH ( nullable = true ),
	update_time varchar WITH ( nullable = true )
 )                                                       
 WITH (                                                  
    number_of_replicas = 3,                              
    partition_by_hash_buckets = 20,                      
    partition_by_hash_columns = ARRAY['grade','id']
 );
-- insert data
    insert into kudu.test.test_fact values('AA1','a1','99.0','stu1','12','my school','hari','2018-01-18 14:00:53');
    insert into kudu.test.test_fact values('AA2','a1','98.5','stu1','13','my school','hari','2018-01-18 14:00:53');
    insert into kudu.test.test_fact values('AA3','a2','96.5','stu1','14','my school','hari','2018-01-18 14:00:53');
    insert into kudu.test.test_fact values('AA4','a2','99.1','stu1','11','my school','hari','2018-01-18 14:00:53');
    insert into kudu.test.test_fact values('AA5','a4','88.5','stu1','12','my school','hari','2018-01-18 14:00:53');
    insert into kudu.test.test_fact values('AA6','a1','70.5','stu1','13','my school','hari','2018-01-18 14:00:53');
    insert into kudu.test.test_fact values('AA7','a4','90.5','stu1','10','my school','hari','2018-01-18 14:00:53');

-- The result is a field mismatch.Notice the teacher field in the query result,the value of the school field is displayed.
select t1.id as id ,t1.name as name,t1.score as score,t1.age as age,t1.teacher as teacher,t1.update_time as update_time
 from kudu.test.test_fact t1
  where (t1.grade = 'a1') and t1.age in (select age from kudu.test.test_fact) and t1.name is not null;

 -- When the school was checked out, the result showed normal, no dislocation. 
select t1.id as id ,t1.name as name,t1.score as score,t1.age as age,t1.school as school,t1.teacher as teacher,t1.update_time as update_time
 from kudu.test.test_fact t1
  where (t1.grade = 'a1') and t1.age in (select age from kudu.test.test_fact) and t1.name is not null;

image

### PR DESCRIPTION
**Pull request by Praveen2112**
Fixes #3170, Fixes #2963 
Supersedes https://github.com/prestosql/presto/pull/3193

### CODE REVIEW

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduRecordCursor.java [hunk 1]
```diff
@@ -14,19 +14,26 @@
 package io.prestosql.plugin.kudu;
```
Comment (kokosing) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduRecordCursor.java]: Can you please add a tests where incorrect column mapping was exposed?

File: presto-kudu/src/test/java/io/prestosql/plugin/kudu/TestKuduDistributedQueries.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    @Override
+    public void testDescribeOutputNamedAndUnnamed()
+    {
+        // this connector uses a non-canonical type for varchar columns in tpch
+    }
+
+    @Override
+    public void testCommentTable()
+    {
+        // Kudu connector currently does not support comment on table
```
Comment (kokosing) [inline presto-kudu/src/test/java/io/prestosql/plugin/kudu/TestKuduDistributedQueries.java]: this comment is not needed, the code below is self explanatory

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduRecordCursor.java (hunk 1 above)
Comment (Praveen2112) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduRecordCursor.java]: AbstractTestQuery#testCorrelatedJoin, testTableSampleWithFiltering, testSelectAllFromOuterScopeTable exposes this incorrect table mapping. `TestKuduDistributedQueries`  captures most of the cases where incorrect column mapping was exposed

**Comment by kokosing** [comment]
Please ping me when automation passes.

**Pull request merged**